bayem/json_io.py

```diff
@@ -0,0 +1,128 @@
+import json
+
+import numpy as np
+
+from .distributions import MVN, Gamma
+from .vba import VBResult
+
+
+class BayemEncoder(json.JSONEncoder):
+    """
+    Usage:
+        string = json.dumps(obj, cls=bayem.BayemEncoder, ...)
+        with open(...) as f:
+            json.dump(obj, f cls=bayem.BayemEncoder, ...)
+
+    Details:
+
+    Out of the box, JSON can serialize
+        dict, list, tuple, str, int, float, True/False, None
+
+    To make our custom classes JSON serializable, we subclass from JSONEncoder
+    and overwrite its `default` method to somehow represent our class with the
+    types provided above.
+
+    The idea is to serialize our custom classes (and numpy...) as a dict
+    containing:
+        key: unique string to represent the classe -- this helps us to indentify
+             the class when _de_serializing in `bayem.bayem_hook` below
+        value: some json-serializable entries -- obj.__dict__ contains all
+               members class members and is not optimal, but very convenient.
+    https://stackoverflow.com/questions/3768895/how-to-make-a-class-json-serializable
+
+    Note that the method is called recursively. To serialize MVN we call the
+    `default` method below and see that json representation should contain
+    the __dict__ of members. This dict also contains "mean" of type `np.array`.
+    Thus, `default` is now called on `np.array` which is represented by a list
+    of its values.
+
+    """
+
+    def default(self, obj):
+        """
+        `obj`:
+            python object to serialize
+        return:
+            json (serializeable) reprensentation of `obj`
+        """
+        if isinstance(obj, VBResult):
+            return {"bayem.VBResult": obj.__dict__}
```

Comment:
Could you document what is stored in this __dict__ object, I found this (The namespace supporting arbitrary function attributes), but on a first glanced I'm confused why this not covered by the default.

Reply:
https://docs.python.org/3/library/stdtypes.html#object.__dict__: 
> A dictionary or other mapping object used to store an object’s (writable) attributes.

So it basically contains all the attributes of the object as a dict. But IMO this is well documented in python itself and the explanation does not belong in here. I also added a stackoverflow link above for further reading. But let me know, if you think otherwise. No hard feelings about that.

Reply:
yes, for me it is sufficient have a link for someone not as profound in python as me.